kernel/printk/printk.c

```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+
+	/* Offset sequence numbers by the number of existing bootloader msgs 
+	   NOTE: Functions above handle if the first seq becomes > than these */
+	if ( log_cb->syslog_seq != new_cb->log_first_seq )
+		new_cb->syslog_seq = log_cb->syslog_seq + new_cb->log_next_seq;
+	if ( log_cb->console_seq != new_cb->log_first_seq )
+		new_cb->console_seq = log_cb->console_seq + new_cb->log_next_seq;
+	if ( log_cb->clear_seq != new_cb->log_first_seq )
+		new_cb->clear_seq = log_cb->clear_seq + new_cb->log_next_seq;
+
+	/* Copy kernel specific data into the new/shared control block. */
+	new_cb->syslog_prev = log_cb->syslog_prev;
+	new_cb->syslog_partial = log_cb->syslog_partial;
+	new_cb->console_prev = log_cb->console_prev;
+
+	/*
+	   Copy existing entries one at a time into the external buffer space.
+	   The new CB will incrementally update as each kernel entry is added.
+	   The existing CB will not be modified.
+	 */
+	for ( currkmsg_seq = log_cb->log_first_seq,
+			currkmsg_idx = log_cb->log_first_idx;
+			currkmsg_seq <= log_cb->log_next_seq &&
+			currkmsg_idx < log_cb->log_length;
+			currkmsg_idx = log_next ( currkmsg_idx ),
+			currkmsg_seq++ )
+	{
+		currkmsg =
+			log_from_idx_and_base ( log_cb->log_buf, currkmsg_idx );
+
+		/* If necessary, free bl entries to fit kernel entries */
+		while (new_cb->log_first_seq < new_cb->log_next_seq)
+		{
+			u32 free;
+
+			if (new_cb->log_next_idx > new_cb->log_first_idx)
+				free = max(new_cb->log_length -
+						new_cb->log_next_idx,
+						new_cb->log_first_idx);
+			else
+				free = new_cb->log_first_idx -
+					new_cb->log_next_idx;
+
+			if (free > currkmsg->len + sizeof(struct printk_log))
+				break;
+
+			/* Drop oldest messages until there is enough space */
+			new_cb->log_first_idx =
+				log_next_idx_from_idx (
+						new_cb->log_buf,
+						new_cb->log_first_idx );
+			new_cb->log_first_seq++;
+		}
+
+		nextblmsg =
+			(struct printk_log *)
+			( new_cb->log_buf + new_cb->log_next_idx );
+
+		// Check if the next record will fit along with a wrap record
+		if ( new_cb->log_next_idx + currkmsg->len + sizeof(struct printk_log) >=
+				new_cb->log_length )
+		{
+			memset ( nextblmsg, 0, sizeof(struct printk_log) );
+			nextblmsg = (struct printk_log*)new_cb->log_buf;
+			new_cb->log_next_idx = 0;
+		}
+
+		// Copy kernel message to bootloader buffer
+		memcpy ( nextblmsg, currkmsg, currkmsg->len );
+
+		/* Before moving to the next kernel entry, copy adjusted indices */
+		if ( currkmsg_seq == log_cb->log_first_seq )
+			offset_idx = new_cb->log_first_seq;
+		else
+			offset_idx = new_cb->log_next_idx;
+
+		if ( currkmsg_seq == log_cb->syslog_seq )
+			new_cb->syslog_idx = offset_idx;
+		if ( currkmsg_seq == log_cb->console_seq )
+			new_cb->console_idx = offset_idx;
+		if ( currkmsg_seq == log_cb->clear_seq )
+			new_cb->clear_idx = offset_idx;
+
+		// Update next offset and sequence
+		// Note: avoid incrementing too far by excluding the last pass
```

Comment:
It's not clear to me at a first glance what this note is intending to communicate.  Is there a potential bug here?  Is it explaining the logic behind the conditional that follows?